wire/msg.go

```diff
@@ -42,40 +53,35 @@ type (
 		// Msg contained in this Envelope. Not embedded so Envelope doesn't implement Msg.
 		Msg Msg
 	}
-)
 
-// Encode encodes an Envelope into an io.Writer.
-func (env *Envelope) Encode(w io.Writer) error {
-	if err := perunio.Encode(w, env.Sender, env.Recipient); err != nil {
-		return err
+	// EnvelopeSerializer serializes/deserializes envelopes into/from streams.
+	EnvelopeSerializer interface {
+		Encode(w io.Writer, env Envelope) error
+		Decode(r io.Reader) (Envelope, error)
 	}
-	return Encode(env.Msg, w)
+)
+
+// EncodeEnvelope serializes envelopes into the writer, using the global
+// envelope serialzer instance.
+func EncodeEnvelope(w io.Writer, env Envelope) error {
+	return envelopeSerializer.Encode(w, env)
 }
 
-// Decode decodes an Envelope from an io.Reader.
-func (env *Envelope) Decode(r io.Reader) (err error) {
-	env.Sender = NewAddress()
-	if err = perunio.Decode(r, env.Sender); err != nil {
-		return err
-	}
-	env.Recipient = NewAddress()
-	if err = perunio.Decode(r, env.Recipient); err != nil {
-		return err
-	}
-	env.Msg, err = Decode(r)
-	return err
+// DecodeEnvelope deserializes envelopes from streams, using the global
+// envelope serialzer instance.
+func DecodeEnvelope(r io.Reader) (Envelope, error) {
+	return envelopeSerializer.Decode(r)
 }
 
-// Encode encodes a message into an io.Writer. It also encodes the
-// message type whereas the Msg.Encode implementation is assumed not to write
-// the type.
-func Encode(msg Msg, w io.Writer) error {
+// EncodeMsg encodes a message into an io.Writer. It also encodes the message
+// type whereas the Msg.Encode implementation is assumed not to write the type.
+func EncodeMsg(msg Msg, w io.Writer) error {
```

Comment:
I agree to the point that, these are artifacts of perunio encoding and should not be exported.

But, these functions and the associated Decoders map is still located in `wire` package. Hence, these functions and the have to be exported, so that they can be used in `wire/perunio`.

Moving the decoders map and these functions to  `perunio` in itself will involve changes in many places. Hence it is described as a separate step (step 6), in #233.